components/eam/src/physics/cam/micro_p3_interface.F90

```diff
@@ -984,38 +986,41 @@ subroutine micro_p3_tend(state, ptend, dtime, pbuf)
-    rain(:ncol,:pver)        = calculate_drymmr_from_wetmmr(ncol, pver, state%q(:,:,ixrain),    qv_wet_in)
-    numrain(:ncol,:pver)     = calculate_drymmr_from_wetmmr(ncol, pver, state%q(:,:,ixnumrain), qv_wet_in)
-    ice(:ncol,:pver)         = calculate_drymmr_from_wetmmr(ncol, pver, state%q(:,:,ixcldice),  qv_wet_in)
-    qm(:ncol,:pver)          = calculate_drymmr_from_wetmmr(ncol, pver, state%q(:,:,ixcldrim),  qv_wet_in) !Aaron, changed ixqm to ixcldrim to match Kai's code
-    numice(:ncol,:pver)      = calculate_drymmr_from_wetmmr(ncol, pver, state%q(:,:,ixnumice),  qv_wet_in)
-    rimvol(:ncol,:pver)      = calculate_drymmr_from_wetmmr(ncol, pver, state%q(:,:,ixrimvol),  qv_wet_in)
+
+    ratio_local(:ncol,:pver) = state%pdel(:ncol,:pver)/state%pdeldry(:ncol,:pver)
+
+    qv_dry     (:ncol,:pver) = state%q(:ncol,:pver,         1) *ratio_local(:ncol,:pver)
+    cldliq     (:ncol,:pver) = state%q(:ncol,:pver,  ixcldliq) *ratio_local(:ncol,:pver)
+    numliq     (:ncol,:pver) = state%q(:ncol,:pver,  ixnumliq) *ratio_local(:ncol,:pver)
+    rain       (:ncol,:pver) = state%q(:ncol,:pver,    ixrain) *ratio_local(:ncol,:pver)
+    numrain    (:ncol,:pver) = state%q(:ncol,:pver, ixnumrain) *ratio_local(:ncol,:pver)
+    ice        (:ncol,:pver) = state%q(:ncol,:pver,  ixcldice) *ratio_local(:ncol,:pver)
+    !Aaron, changed ixqm to ixcldrim to match Kai's code
+    qm         (:ncol,:pver) = state%q(:ncol,:pver,  ixcldrim) *ratio_local(:ncol,:pver)
+    numice     (:ncol,:pver) = state%q(:ncol,:pver,  ixnumice) *ratio_local(:ncol,:pver)
+    rimvol     (:ncol,:pver) = state%q(:ncol,:pver,  ixrimvol) *ratio_local(:ncol,:pver)
+
+    qv_prev_dry(:ncol,:pver) = qv_prev_wet(:ncol,:pver)        *ratio_local(:ncol,:pver)
 
     ! COMPUTE GEOMETRIC THICKNESS OF GRID & CONVERT T TO POTENTIAL TEMPERATURE
     !==============
     ! TODO: Create a general function to calculate Exner's formula that can be
     ! used by all parameterizations, such as P3 and SHOC.
     ! This would take a bit more work, so we have decided to delay this task
     ! until a later stage of code cleanup.
-    inv_exner(:ncol,:pver) = 1._rtype/((state%pmiddry(:ncol,:pver)*1.e-5_rtype)**(rair*inv_cp))
+
+    inv_exner(:ncol,:pver) = 1._rtype/((state%pmid(:ncol,:pver)*1.e-5_rtype)**(rair*inv_cp))
     do icol = 1,ncol
        do k = 1,pver
           ! Note, there is a state%zi variable that could be used to calculate
           ! dz, but that is in a wet coordinate frame rather than dry.  Now that
           ! P3 is using dry MMR we instead calculated dz using virtual
           ! temperature and pressure.
-          T_virtual  = state%t(icol,k) * (1.0 + qv_dry(icol,k)*(1.0*mwdry/mwh2o - 1.0))
-          dz(icol,k) = (rair/gravit) * state%pdeldry(icol,k) * T_virtual / state%pmiddry(icol,k) 
+
+          T_virtual  = state%t(icol,k) * (1.0 + state%q(icol,k,1)*(mwdry/mwh2o - 1.0))
+          dz(icol,k) = (state%zi(icol,k) - state%zi(icol,k+1))/gravit
```

Comment:
Since `T_virtual` is not being used to calculate `dz`, it doesn't seem to be used at all, so I think you should just remove `T_virtual` completely.